sw/device/lib/base/hardened.h

```diff
@@ -304,4 +304,158 @@ inline void barrier32(uint32_t val) { asm volatile("" ::"r"(val)); }
  */
 inline void barrierw(uintptr_t val) { asm volatile("" ::"r"(val)); }
 
+/**
+ * A constant-time, 32-bit boolean value.
+ *
+ * Values of this type MUST be either all zero bits or all one bits.
+ */
+typedef uint32_t ct_bool32_t;
+
+// NOTE: the functions below are taken from BoringSSL, which includes proofs of
+// correctness by Adam Langley.
+// https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/internal.h#262
+
+/**
+ * Performs constant-time signed comparison to zero.
+ *
+ * Returns whether `a < 0`, as a constant-time boolean.
+ *
+ * @return `a < 0`.
+ */
+inline ct_bool32_t ct_sltz32(int32_t a) {
+  // Arithmetic left shift!
+  return (uint32_t)(a >> 31);
+}
+
+/**
+ * Performs constant-time unsigned ascending comparison.
+ *
+ * Returns `a < b` as a constant-time boolean.
+ *
+ * @return `a < b`.
+ */
+inline ct_bool32_t ct_sltu32(uint32_t a, uint32_t b) {
+  return ct_sltz32(a ^ ((a ^ b) | ((a - b) ^ a)));
+}
+
+/**
+ * Performs constant-time unsigned descending comparison.
+ *
+ * Returns `a > b` as a constant-time boolean.
+ *
+ * @return `a > b`.
+ */
+inline ct_bool32_t ct_sgeu32(uint32_t a, uint32_t b) {
+  return ~ct_sltu32(b, a);
```

Comment:
The name here (`geu`) doesn't seem to match the comment (`gtu`).

Also `~ct_sltu32(b, a)` is `¬(b < a)` which is equivalent to `b >= a`. That seems wrong in either case.

Reply:
Comment should have been `gtu`. Either way, I'm dropping this one, since it's not a useful primitive.